### PR TITLE
[Android] add wrapper for workload-automation

### PR DESCRIPTION
**Pull request by mdigiorgio**
Workload-automation (WA) already provides many workloads for Android targets.
This is a wrapper for LISA to use WA.

In order to use one of the WA workloads (`wa list workloads` tells you what is
available) you need to implement a <workload__class_name> class in a
<workload_name>.py file stored under libs/utils/android/workloads.

<workload_class_name> inherits Wlauto and has to implement the methods from
Workload and (eventually) Wlauto.

This way, the workload is visible by the Workload class that will list it as
available and will eventually check that the related package is installed on the
Android target.

Signed-off-by: Michele Di Giorgio michele.digiorgio@arm.com

### CODE REVIEW

File: libs/utils/android/wlauto.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            filename. In the latter case the file must be under
+            android/workloads.
+        :type agenda: str
+        """
+        # Check if the specified path is absolute
+        if agenda.startswith('/'):
+            path_to_agenda = agenda
+        else:
+            # Assume only file name
+            path_to_agenda = os.path.join(workloads_dir, agenda)
```
Comment (derkling) [inline libs/utils/android/wlauto.py]: Should be better use a subfolder for agendas?

File: libs/utils/android/wlauto.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+        logging.debug('%14s - Using agenda %s',
+                      'Wlauto',
+                      path_to_agenda)
+
+        # Create temprary output directory for WA
+        tmp_dir = tempfile.mkdtemp()
```
Comment (derkling) [inline libs/utils/android/wlauto.py]: Should we add something like: prefix="lisa_wa_"?

File: libs/utils/android/wlauto.py (hunk 1 above)
Comment (derkling) [inline libs/utils/android/wlauto.py]: Should we use a shorter _agenda?

File: libs/utils/android/wlauto.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        logging.debug('%14s - wa output dir is %s', 'Wlauto', tmp_dir)
+
+        # Execute workload-automation
+        p = sp.Popen(WA_RUN_CMD.format(path_to_agenda, tmp_dir).split(),
+                     stdout=sp.PIPE,
+                     stderr=sp.PIPE)
+        out, err = p.communicate()
+        logging.debug(out)
+        logging.debug(err)
+        return tmp_dir
```
Comment (derkling) [inline libs/utils/android/wlauto.py]: Should we rename this to results_dir... at the end is where WA preduces it's output, isn't it?

File: libs/utils/android/wlauto.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+    def generate_yaml_agenda(self, agenda):
+        """
+        Converts a dictionary into a YAML file which should used as the agenda
```
Comment (derkling) [inline libs/utils/android/wlauto.py]: nit: which should BE

File: libs/utils/android/wlauto.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        for running workload-automation benchmarks.
+
+        The YAML file is put under the result directory specified in the test
```
Comment (derkling) [inline libs/utils/android/wlauto.py]: resultS directory

File: libs/utils/android/wlauto.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        environment.
+
+        :param agenda: Dictionary to be converted into YAML file
+        :type agenda: dict
+        """
+
+        agenda_file = os.path.join(self.te.res_dir, 'agenda.yaml')
```
Comment (derkling) [inline libs/utils/android/wlauto.py]: again, just _agenda?

File: libs/utils/android/wlauto.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
```
Comment (derkling) [inline libs/utils/android/wlauto.py]: Should we:
```python
if not isinstance(agenda, dict):
    raise ValueError('Specified agenda must be a dictionary')
```

File: libs/utils/android/workload.py [hunk 8]
```diff
@@ -92,7 +92,7 @@ def availables(target):
                     module_name = module.__name__
                     # Check if a package is required and is available on target
                     aw = Workload._AW(module_name, module, class_name, handler)
-                    if (Workload._is_available(target, aw)):
```
Comment (derkling) [inline libs/utils/android/workload.py]: Was this broken since before this patch?!?
In that case we shoud better post this as a separate fix patch.

File: libs/utils/android/workloads/camerapreview.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+from subprocess import Popen, PIPE
+from android import Screen, Workload
+from time import sleep
+from devlib.utils.android import adb_command
+
+import logging
+
+GET_FRAME_STATS_CMD = 'shell dumpsys gfxinfo {} > {}'
+
+class CameraPreview(Workload):
```
Comment (derkling) [inline libs/utils/android/workloads/camerapreview.py]: This is not a Wlauto workload... better add it as a separate patch...

File: libs/utils/android/workloads/camerapreview.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+    def run(self, exp_dir, preview_duration_s, collect=''):
+
+        # Initialize energy meter results
+        nrg_data, nrg_file = None, None
+
+        # Unlock device screen (assume no password required)
+        self.target.execute('input keyevent 82')
+        # Press Back button to be sure we run the video from the start
+        self.target.execute('input keyevent KEYCODE_BACK')
```
Comment (derkling) [inline libs/utils/android/workloads/camerapreview.py]: There is a new Android::System module available... it would be great if we can move these previous two commands to be a standard API of that module. Something like:
```
android.system.unlock_screen()
android.system.press_back_button()
```

Comment (derkling) [inline libs/utils/android/workloads/camerapreview.py]: Maybe you can add this code as is right now... and than add a patch which adds System functionalities followed by a patch which refactors all the libs/utils/android/ modules which make use of these calls.
For example, there are similar calls in:
https://github.com/ARM-software/lisa/blob/master/libs/utils/android/workloads/jankbench.py#L83

As well as some other APIs:

```python
self.target.execute('am force-stop com.android.benchmark')
self.target.execute('pm clear com.android.benchmark')
```

Which should be better become a:
```Python
System.force_stop(apk_name, clear=True)
```

File: libs/utils/android/workloads/camerapreview.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+        # Force screen in LANDSCAPE mode
+        Screen.set_orientation(self.target, portrait=False)
+
+        # Start CameraPreview video on the target device
+        self.target.execute('am start -n {}/{}'.format(self.package,
+                                                       self.activity))
```
Comment (derkling) [inline libs/utils/android/workloads/camerapreview.py]: Another candidate for System API

```
System.start(apk_name, activity_name)
```

File: libs/utils/android/workloads/camerapreview.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        # Allow the activity to start
+        sleep(3)
+
+        # Reset framestats collection
+        self.target.execute('dumpsys gfxinfo {} reset'.format(self.package))
```
Comment (derkling) [inline libs/utils/android/workloads/camerapreview.py]: Yet another candidate...
```python
System.collect_framestats(apk_name)
```

File: libs/utils/android/workloads/camerapreview.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+        # Stop energy collection
+        if 'energy' in collect and self.te.emeter:
+            nrg_data, nrg_file = self.te.emeter.report(exp_dir)
+            logging.info("Estimated energy: %7.3f", float(nrg_data['BAT']))
+
+        # Get frame stats
+        db_file = os.path.join(exp_dir, "framestats.txt")
+        adb_command(self.target.adb_name,
+                    GET_FRAME_STATS_CMD.format(self.package, db_file))
```
Comment (derkling) [inline libs/utils/android/workloads/camerapreview.py]: Why this adb_command call?
However... another candidate to pair with the previous System.collect_framestats(apk_name)...

File: libs/utils/android/workloads/camerapreview.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+        nrg_data, nrg_file = None, None
+
+        # Unlock device screen (assume no password required)
+        self.target.execute('input keyevent 82')
+        # Press Back button to be sure we run the video from the start
+        self.target.execute('input keyevent KEYCODE_BACK')
+
+        # Force screen in LANDSCAPE mode
+        Screen.set_orientation(self.target, portrait=False)
+
```
Comment (derkling) [inline libs/utils/android/workloads/camerapreview.py]: Here we need also to do two additional thinks:
1. use max backlight (Screen API)
2. set AIRPLANE mode (System API)

File: libs/utils/android/wlauto.py (hunk 3 above)
Comment (mdigiorgio) [inline libs/utils/android/wlauto.py]: Yes, it is

File: libs/utils/android/workloads/camerarecord.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+        # Initialize energy meter results
+        nrg_data, nrg_file = None, None
+
+        # Setting agenda_yaml properly
+        if type(agenda) is str:
+            agenda_yaml = agenda
+        else:
+            # Convert dictionary to YAML file
+            agenda_yaml = self.generate_yaml_agenda(agenda)
```
Comment (derkling) [inline libs/utils/android/workloads/camerarecord.py]: Overall I would prefer something like:
```python
_agenda = agenda
if type(agenda) is str:
    _agenda = self.generate_yaml_agenda(agenda)
```

I.e remove if conditions by using a default assignment...

File: libs/utils/android/workloads/camerarecord.py [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+        # Start energy collection
+        if 'energy' in collect and self.te.emeter:
+            self.te.emeter.reset()
+
+        wa_out_dir = Wlauto.wa_run(agenda_yaml)
+
+        # Stop energy collection
+        if 'energy' in collect and self.te.emeter:
+            nrg_data, nrg_file = self.te.emeter.report(exp_dir)
+            logging.info("Estimated energy: %7.3f", float(nrg_data['BAT']))
```
Comment (derkling) [inline libs/utils/android/workloads/camerarecord.py]: Here we still have that hardcoded BAT lable... have we never decided how to make that more generic?

File: libs/utils/android/wlauto.py (hunk 7 above)
Comment (mdigiorgio) [inline libs/utils/android/wlauto.py]: Yes better :)

File: libs/utils/android/workloads/camerarecord.yaml [hunk 17]
```diff
@@ -0,0 +1,7 @@
+config:
+    device: generic_android
+workloads:
+    - id: b01
```
Comment (derkling) [inline libs/utils/android/workloads/camerarecord.yaml]: Maybe use a more meaningful name?
I guess this is reported in logs as well... something like "lisa_camerarecord_20" ?

File: libs/utils/android/workloads/rtapp.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+from android import Wlauto
+from android import Workload
+
+import os
+import glob
+import logging
+
+from shutil import copyfile
+
+class RTApp(Wlauto):
```
Comment (derkling) [inline libs/utils/android/workloads/rtapp.py]: Overall same comments as per the camerarecord case...

One more consideration, should we better name these workloads to identify that they are provided via WA?
This means:
1. rename the workload to something like: WaRTapp, WaCameRarecord
2. rename the file as well: wa_rtapp.py, wa_camerarecord.py

File: libs/utils/android/workloads/vellamo.py [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+import os
+import collections
+import copy
+import logging
+
+VELLAMO_DB_PATH = '/data/data/com.quicinc.vellamo/files'
+VELLAMO_SCORE_NAME = 'chapterscores.json'
+VELLAMO_TESTS = ['BROWSER', 'METAL', 'MULTI']
+
+class Vellamo(Workload):
```
Comment (derkling) [inline libs/utils/android/workloads/vellamo.py]: Post this on a separate patch, maybe in the same patchset of CameraPreview.
Let's have a separate PR for these non WA related WL... they are more easy to merge ;-)

File: libs/utils/android/workload.py (hunk 8 above)
Comment (mdigiorgio) [inline libs/utils/android/workload.py]: No, as far as I remember we introduced this for making WA wrapper work

File: libs/utils/android/workloads/vellamo.py [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+
+        if test_name.upper() not in VELLAMO_TESTS:
+            raise ValueError('Vellamo workload [%s] not supported', test_name)
+
+        # Set parameter depending on test
+        self.logger.debug('Start Vellamo Benchmark [%s]', test_name)
+        test_cmd = ''
+        sleep_time = 0
+        if test_name.upper() == 'BROWSER':
+            test_cmd = 'input tap 970 740'
```
Comment (derkling) [inline libs/utils/android/workloads/vellamo.py]: Any chance to use "percentages of screen resolutions" for these tap?
I have a 2K display to cross-check if we want to give it a try... and maybe we can find a 720p or lower as well...

File: libs/utils/android/workloads/vellamo.py [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+        self.target.execute('input keyevent 82')
+        # Start Vellamo on the target device
+        self.target.execute('pm clear {}'.format(self.package))
+        self.target.execute('am start -W -n {}/{}'.format(self.package,
+                                                          self.activity))
+        # Wait few seconds to make sure the app is loaded
+        sleep(1)
+
+        # Accept EULA
+        self.target.execute('input tap 864 1646')
```
Comment (derkling) [inline libs/utils/android/workloads/vellamo.py]: And a new API similar to:
```python
System.tap(x_perc, y_perc)
```
would make thinks :100: times more cool

File: libs/utils/android/workloads/vellamo.py [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+
+        # Start measuring energy
+        if 'energy' in collect and self.te.emeter:
+            self.te.emeter.reset()
+
+        # Run Workload
+        self.target.execute(test_cmd)
+
+        # Skip instructions
+        self.target.execute('input swipe 900 1000 200 1000 100')
```
Comment (derkling) [inline libs/utils/android/workloads/vellamo.py]: What about?
```python
System.hswipe(x_perc, y_lower_perc, y_upper_perc, swipe_up=True)
```

and the companion
```python
System.hswipe(y_perc, x_left_perc, x_right_perc, swipe_right=True)
```

File: libs/utils/android/workloads/vellamo.py [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+
+        # Wait for completion
+        sleep(sleep_time*60)
+
+        # Stop measuring energy
+        if 'energy' in collect and self.te.emeter:
+            nrg_data, nrg_file = self.te.emeter.report(exp_dir)
+            self.logger.info('%s - Estimated energy: %7.3f',
+                             test_name,
+                             float(nrg_data['BAT']))
```
Comment (derkling) [inline libs/utils/android/workloads/vellamo.py]: To Be Fixed...

Comment (derkling) [inline libs/utils/android/workloads/vellamo.py]: Do you think it could be sufficient, at least for the time being, to expose the channel name as a run() method parameter?!? At the end the EMeter is configured by the caller... thus:
1. if we specify a channel: that means we want to collect energy and we know also from which channel.
```python
def wload_run(self, exp_dir, test_name, collect='', emeter_channel=None):
    #....
    if 'energy' in collect and self.te.emeter and emeter_channel:
        #....
```

File: libs/utils/android/workloads/videopreview.py [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+from subprocess import Popen, PIPE
+from android import Screen, Workload
+from time import sleep
+from devlib.utils.android import adb_command
+
+import logging
+
+GET_FRAME_STATS_CMD = 'shell dumpsys gfxinfo {} > {}'
+
+class VideoPreview(Workload):
```
Comment (derkling) [inline libs/utils/android/workloads/videopreview.py]: Move in the same PR for the non WA workloads...

File: libs/utils/android/workloads/youtube.py [hunk 25]
```diff
@@ -56,6 +56,11 @@ def run(self, exp_dir, video_url, video_duration_s, collect=''):
         # Force screen in LANDSCAPE mode
         Screen.set_orientation(self.target, portrait=False)
 
+        self.target.execute('monkey -p {} 1'.format(self.package))
```
Comment (derkling) [inline libs/utils/android/workloads/youtube.py]: These seems more a fix to be posted in a separate patch....
BTW: what is that?!?

File: libs/utils/android/workloads/camerapreview.py (hunk 13 above)
Comment (mdigiorgio) [inline libs/utils/android/workloads/camerapreview.py]: Because I want to save the db_file locally. It is similar to what you wrote in `workload.py`, it is was already available in `devlib`

File: libs/utils/android/workloads/camerarecord.py (hunk 16 above)
Comment (mdigiorgio) [inline libs/utils/android/workloads/camerarecord.py]: No, we still haven't discuss that

File: libs/utils/android/workloads/vellamo.py (hunk 20 above)
Comment (mdigiorgio) [inline libs/utils/android/workloads/vellamo.py]: Not sure, this is a button. I don't know where it will be placed with different screens...

File: libs/utils/android/workloads/vellamo.py (hunk 23 above)
Comment (mdigiorgio) [inline libs/utils/android/workloads/vellamo.py]: Yes, we can do this

File: libs/utils/android/workloads/youtube.py (hunk 25 above)
Comment (mdigiorgio) [inline libs/utils/android/workloads/youtube.py]: This was needed to just start YouTube without any video if I well remember

File: libs/utils/android/workload.py (hunk 8 above)
Comment (derkling) [inline libs/utils/android/workload.py]: Still I don't remember the purpose and it's not easy to figure it out from the code.
We should:
1. use better var names, let's redefined:
```python
AndroidWorkload = namedtuple('AndroidWorkload',  ['module_name', 'module', 'class_name', 'ctor'])
```
2. add a better comment which explain why we need that check

**Comment by derkling** [comment]
This patch needs to be rebased on master and split in a series providing:
- the base Wlauto class which extends the Workload base class
- one patch for each new workload added

**Comment by mdigiorgio** [comment]
Rebased and split as requested :)
